CLI: Show fixes by default in `ruff check`

Requires #7350 for opt-out of display.

`ruff check` should display fixes that Ruff can apply. Currently this is only possible for the whole file with `--diff`.

Some fixes may modify content across a file, e.g. when renaming a binding. We should limit the display of fixes to roughly match the range of source context displayed (#7349). If we do not display the entire fix, we should include a notice of the number of additional lines that will be modified if the fix is applied. Naively, we may be able to implement this by only selecting a portion of the fix. In a more comprehensive implementation, we may need to add a new diagnostic range attribute for this display and address it in each problem case.

> Clippy has a nice example user experience of showing fixes:
> 
> 
> ```
> error[E0308]: mismatched types
> --> crates/ruff_python_semantic/src/model.rs:224:21
>     |
> 224 |             source: &self.stmt_id,
>     |                     ^^^^^^^^^^^^^ expected `Option<NodeId>`, found `&Option<NodeId>`
>     |
>     = note:   expected enum `std::option::Option<node::NodeId>`
>             found reference `&std::option::Option<node::NodeId>`
> help: consider removing the borrow
>     |
> 224 -             source: &self.stmt_id,
> 224 +             source: self.stmt_id,
>     |
> 
> For more information about this error, try `rustc --explain E0308`.
> ```

--- a/crates/ruff/src/printer.rs
+++ b/crates/ruff/src/printer.rs
@@ -30,8 +30,6 @@ bitflags! {
         const SHOW_VIOLATIONS = 1 << 0;
         /// Whether to show a summary of the fixed violations when emitting diagnostics.
         const SHOW_FIX_SUMMARY = 1 << 1;
-        /// Whether to show a diff of each fixed violation when emitting diagnostics.
-        const SHOW_FIX_DIFF = 1 << 2;
     }
 }
 
@@ -260,7 +258,7 @@ impl Printer {
             OutputFormat::Concise | OutputFormat::Full => {
                 TextEmitter::default()
                     .with_show_fix_status(show_fix_status(self.fix_mode, fixables.as_ref()))
-                    .with_show_fix_diff(self.flags.intersects(Flags::SHOW_FIX_DIFF))
+                    .with_show_fix_diff(self.format == OutputFormat::Full && preview)
                     .with_show_source(self.format == OutputFormat::Full)
                     .with_fix_applicability(self.unsafe_fixes.required_applicability())
                     .with_preview(preview)

--- a/crates/ruff/src/printer.rs
+++ b/crates/ruff/src/printer.rs
@@ -262,7 +262,7 @@ impl Printer {
                     .with_show_fix_status(show_fix_status(self.fix_mode, fixables.as_ref()))
                     .with_show_fix_diff(self.flags.intersects(Flags::SHOW_FIX_DIFF))
                     .with_show_source(self.format == OutputFormat::Full)
-                    .with_unsafe_fixes(self.unsafe_fixes)
+                    .with_fix_applicability(self.unsafe_fixes.required_applicability())
                     .with_preview(preview)
                     .emit(writer, &diagnostics.inner, &context)?;
 
@@ -464,7 +464,7 @@ impl Printer {
             TextEmitter::default()
                 .with_show_fix_status(show_fix_status(self.fix_mode, fixables.as_ref()))
                 .with_show_source(preview)
-                .with_unsafe_fixes(self.unsafe_fixes)
+                .with_fix_applicability(self.unsafe_fixes.required_applicability())
                 .emit(writer, &diagnostics.inner, &context)?;
         }
         writer.flush()?;

--- a/crates/ruff/tests/lint.rs
+++ b/crates/ruff/tests/lint.rs
@@ -5830,3 +5830,33 @@ nested_optional: Optional[Optional[Optional[str]]] = None
     ",
     );
 }
+
+#[test]
+fn show_fixes_in_full_output_with_preview_enabled() {
+    assert_cmd_snapshot!(
+        Command::new(get_cargo_bin(BIN_NAME))
+            .args(["check", "--no-cache", "--output-format", "full"])
+            .args(["--select", "F401"])
+            .arg("--preview")
+            .arg("-")
+            .pass_stdin("import math"),
+        @r"
+    success: false
+    exit_code: 1
+    ----- stdout -----
+    F401 [*] `math` imported but unused
+     --> -:1:8
+      |
+    1 | import math
+      |        ^^^^
+      |
+    help: Remove unused import: `math`
+      - import math
+
+    Found 1 error.
+    [*] 1 fixable with the `--fix` option.
+
+    ----- stderr -----
+    ",
+    );
+}

--- a/crates/ruff_db/src/diagnostic/mod.rs
+++ b/crates/ruff_db/src/diagnostic/mod.rs
@@ -349,6 +349,13 @@ impl Diagnostic {
         self.fix().is_some()
     }
 
+    /// Returns `true` if the diagnostic is [`fixable`](Diagnostic::fixable) and applies at the
+    /// configured applicability level.
+    pub fn fix_applies(&self, config: &DisplayDiagnosticConfig) -> bool {
+        self.fix()
+            .is_some_and(|fix| fix.applies(config.fix_applicability))
+    }
+
     /// Returns the offset of the parent statement for this diagnostic if it exists.
     ///
     /// This is primarily used for checking noqa/secondary code suppressions.

--- a/crates/ruff_db/src/diagnostic/mod.rs
+++ b/crates/ruff_db/src/diagnostic/mod.rs
@@ -351,7 +351,7 @@ impl Diagnostic {
 
     /// Returns `true` if the diagnostic is [`fixable`](Diagnostic::fixable) and applies at the
     /// configured applicability level.
-    pub fn fix_applies(&self, config: &DisplayDiagnosticConfig) -> bool {
+    pub fn is_fix_applicable(&self, config: &DisplayDiagnosticConfig) -> bool {
         self.fix()
             .is_some_and(|fix| fix.applies(config.fix_applicability))
     }

--- a/crates/ruff_db/src/diagnostic/mod.rs
+++ b/crates/ruff_db/src/diagnostic/mod.rs
@@ -351,7 +351,7 @@ impl Diagnostic {
 
     /// Returns `true` if the diagnostic is [`fixable`](Diagnostic::fixable) and applies at the
     /// configured applicability level.
-    pub fn is_fix_applicable(&self, config: &DisplayDiagnosticConfig) -> bool {
+    pub fn has_applicable_fix(&self, config: &DisplayDiagnosticConfig) -> bool {
         self.fix()
             .is_some_and(|fix| fix.applies(config.fix_applicability))
     }

--- a/crates/ruff_db/src/diagnostic/render.rs
+++ b/crates/ruff_db/src/diagnostic/render.rs
@@ -254,7 +254,7 @@ impl<'a> ResolvedDiagnostic<'a> {
             id,
             message: diag.inner.message.as_str().to_string(),
             annotations,
-            is_fixable: diag.is_fix_applicable(config),
+            is_fixable: diag.has_applicable_fix(config),
         }
     }
 

--- a/crates/ruff_db/src/diagnostic/render.rs
+++ b/crates/ruff_db/src/diagnostic/render.rs
@@ -254,9 +254,7 @@ impl<'a> ResolvedDiagnostic<'a> {
             id,
             message: diag.inner.message.as_str().to_string(),
             annotations,
-            is_fixable: diag
-                .fix()
-                .is_some_and(|fix| fix.applies(config.fix_applicability)),
+            is_fixable: diag.fix_applies(config),
         }
     }
 

--- a/crates/ruff_db/src/diagnostic/render.rs
+++ b/crates/ruff_db/src/diagnostic/render.rs
@@ -254,7 +254,7 @@ impl<'a> ResolvedDiagnostic<'a> {
             id,
             message: diag.inner.message.as_str().to_string(),
             annotations,
-            is_fixable: diag.fix_applies(config),
+            is_fixable: diag.is_fix_applicable(config),
         }
     }
 

--- a/crates/ruff_db/src/diagnostic/render/concise.rs
+++ b/crates/ruff_db/src/diagnostic/render/concise.rs
@@ -77,11 +77,9 @@ impl<'a> ConciseRenderer<'a> {
                     )?;
                 }
                 if self.config.show_fix_status {
-                    if let Some(fix) = diag.fix() {
-                        // Do not display an indicator for inapplicable fixes
-                        if fix.applies(self.config.fix_applicability) {
-                            write!(f, "[{fix}] ", fix = fmt_styled("*", stylesheet.separator))?;
-                        }
+                    // Do not display an indicator for inapplicable fixes
+                    if diag.fix_applies(self.config) {
+                        write!(f, "[{fix}] ", fix = fmt_styled("*", stylesheet.separator))?;
                     }
                 }
             } else {

--- a/crates/ruff_db/src/diagnostic/render/concise.rs
+++ b/crates/ruff_db/src/diagnostic/render/concise.rs
@@ -78,7 +78,7 @@ impl<'a> ConciseRenderer<'a> {
                 }
                 if self.config.show_fix_status {
                     // Do not display an indicator for inapplicable fixes
-                    if diag.is_fix_applicable(self.config) {
+                    if diag.has_applicable_fix(self.config) {
                         write!(f, "[{fix}] ", fix = fmt_styled("*", stylesheet.separator))?;
                     }
                 }

--- a/crates/ruff_db/src/diagnostic/render/concise.rs
+++ b/crates/ruff_db/src/diagnostic/render/concise.rs
@@ -78,7 +78,7 @@ impl<'a> ConciseRenderer<'a> {
                 }
                 if self.config.show_fix_status {
                     // Do not display an indicator for inapplicable fixes
-                    if diag.fix_applies(self.config) {
+                    if diag.is_fix_applicable(self.config) {
                         write!(f, "[{fix}] ", fix = fmt_styled("*", stylesheet.separator))?;
                     }
                 }

--- a/crates/ruff_db/src/diagnostic/render/full.rs
+++ b/crates/ruff_db/src/diagnostic/render/full.rs
@@ -58,7 +58,7 @@ impl<'a> FullRenderer<'a> {
                 writeln!(f, "{}", renderer.render(diag.to_annotate()))?;
             }
 
-            if self.config.show_fix_diff && diag.fix_applies(self.config) {
+            if self.config.show_fix_diff && diag.is_fix_applicable(self.config) {
                 if let Some(diff) = Diff::from_diagnostic(diag, &stylesheet, self.resolver) {
                     write!(f, "{diff}")?;
                 }

--- a/crates/ruff_db/src/diagnostic/render/full.rs
+++ b/crates/ruff_db/src/diagnostic/render/full.rs
@@ -58,7 +58,7 @@ impl<'a> FullRenderer<'a> {
                 writeln!(f, "{}", renderer.render(diag.to_annotate()))?;
             }
 
-            if self.config.show_fix_diff && diag.is_fix_applicable(self.config) {
+            if self.config.show_fix_diff && diag.has_applicable_fix(self.config) {
                 if let Some(diff) = Diff::from_diagnostic(diag, &stylesheet, self.resolver) {
                     write!(f, "{diff}")?;
                 }

--- a/crates/ruff_db/src/diagnostic/render/full.rs
+++ b/crates/ruff_db/src/diagnostic/render/full.rs
@@ -58,7 +58,7 @@ impl<'a> FullRenderer<'a> {
                 writeln!(f, "{}", renderer.render(diag.to_annotate()))?;
             }
 
-            if self.config.show_fix_diff {
+            if self.config.show_fix_diff && diag.fix_applies(self.config) {
                 if let Some(diff) = Diff::from_diagnostic(diag, &stylesheet, self.resolver) {
                     write!(f, "{diff}")?;
                 }

--- a/crates/ruff_db/src/diagnostic/render/full.rs
+++ b/crates/ruff_db/src/diagnostic/render/full.rs
@@ -697,6 +697,8 @@ print()
     fn notebook_output_with_diff() {
         let (mut env, diagnostics) = create_notebook_diagnostics(DiagnosticFormat::Full);
         env.show_fix_diff(true);
+        env.fix_applicability(Applicability::DisplayOnly);
+
         insta::assert_snapshot!(env.render_diagnostics(&diagnostics), @r"
         error[unused-import][*]: `os` imported but unused
          --> notebook.ipynb:cell 1:2:8
@@ -726,7 +728,7 @@ print()
         2 | 
         3 | print('hello world')
 
-        error[unused-variable]: Local variable `x` is assigned to but never used
+        error[unused-variable][*]: Local variable `x` is assigned to but never used
          --> notebook.ipynb:cell 3:4:5
           |
         2 | def foo():
@@ -749,6 +751,7 @@ print()
     fn notebook_output_with_diff_spanning_cells() {
         let (mut env, mut diagnostics) = create_notebook_diagnostics(DiagnosticFormat::Full);
         env.show_fix_diff(true);
+        env.fix_applicability(Applicability::DisplayOnly);
 
         // Move all of the edits from the later diagnostics to the first diagnostic to simulate a
         // single diagnostic with edits in different cells.
@@ -761,7 +764,7 @@ print()
         *fix = Fix::unsafe_edits(edits.remove(0), edits);
 
         insta::assert_snapshot!(env.render(&diagnostic), @r"
-        error[unused-import]: `os` imported but unused
+        error[unused-import][*]: `os` imported but unused
          --> notebook.ipynb:cell 1:2:8
           |
         1 | # cell 1
@@ -924,6 +927,7 @@ line 10
         env.add("example.py", contents);
         env.format(DiagnosticFormat::Full);
         env.show_fix_diff(true);
+        env.fix_applicability(Applicability::DisplayOnly);
 
         let mut diagnostic = env.err().primary("example.py", "3", "3", "label").build();
         diagnostic.help("Start of diff:");
@@ -936,7 +940,7 @@ line 10
         )));
 
         insta::assert_snapshot!(env.render(&diagnostic), @r"
-        error[test-diagnostic]: main diagnostic message
+        error[test-diagnostic][*]: main diagnostic message
          --> example.py:3:1
           |
         1 | line 1

--- a/crates/ruff_linter/src/message/text.rs
+++ b/crates/ruff_linter/src/message/text.rs
@@ -3,9 +3,9 @@ use std::io::Write;
 use ruff_db::diagnostic::{
     Diagnostic, DiagnosticFormat, DisplayDiagnosticConfig, DisplayDiagnostics,
 };
+use ruff_diagnostics::Applicability;
 
 use crate::message::{Emitter, EmitterContext};
-use crate::settings::types::UnsafeFixes;
 
 pub struct TextEmitter {
     config: DisplayDiagnosticConfig,
@@ -46,10 +46,8 @@ impl TextEmitter {
     }
 
     #[must_use]
-    pub fn with_unsafe_fixes(mut self, unsafe_fixes: UnsafeFixes) -> Self {
-        self.config = self
-            .config
-            .fix_applicability(unsafe_fixes.required_applicability());
+    pub fn with_fix_applicability(mut self, applicability: Applicability) -> Self {
+        self.config = self.config.fix_applicability(applicability);
         self
     }
 
@@ -86,13 +84,13 @@ impl Emitter for TextEmitter {
 #[cfg(test)]
 mod tests {
     use insta::assert_snapshot;
+    use ruff_diagnostics::Applicability;
 
     use crate::message::TextEmitter;
     use crate::message::tests::{
         capture_emitter_notebook_output, capture_emitter_output, create_diagnostics,
         create_notebook_diagnostics, create_syntax_error_diagnostics,
     };
-    use crate::settings::types::UnsafeFixes;
 
     #[test]
     fn default() {
@@ -117,7 +115,7 @@ mod tests {
         let mut emitter = TextEmitter::default()
             .with_show_fix_status(true)
             .with_show_source(true)
-            .with_unsafe_fixes(UnsafeFixes::Enabled);
+            .with_fix_applicability(Applicability::Unsafe);
         let content = capture_emitter_output(&mut emitter, &create_diagnostics());
 
         assert_snapshot!(content);
@@ -128,7 +126,7 @@ mod tests {
         let mut emitter = TextEmitter::default()
             .with_show_fix_status(true)
             .with_show_source(true)
-            .with_unsafe_fixes(UnsafeFixes::Enabled);
+            .with_fix_applicability(Applicability::Unsafe);
         let (messages, notebook_indexes) = create_notebook_diagnostics();
         let content = capture_emitter_notebook_output(&mut emitter, &messages, &notebook_indexes);
 

--- a/crates/ruff_linter/src/test.rs
+++ b/crates/ruff_linter/src/test.rs
@@ -447,7 +447,7 @@ pub(crate) fn print_jupyter_messages(
         .with_show_fix_status(true)
         .with_show_fix_diff(true)
         .with_show_source(true)
-        .with_fix_applicability(Applicability::Unsafe)
+        .with_fix_applicability(Applicability::DisplayOnly)
         .emit(
             &mut output,
             diagnostics,
@@ -468,7 +468,7 @@ pub(crate) fn print_messages(diagnostics: &[Diagnostic]) -> String {
         .with_show_fix_status(true)
         .with_show_fix_diff(true)
         .with_show_source(true)
-        .with_fix_applicability(Applicability::Unsafe)
+        .with_fix_applicability(Applicability::DisplayOnly)
         .emit(
             &mut output,
             diagnostics,

--- a/crates/ruff_linter/src/test.rs
+++ b/crates/ruff_linter/src/test.rs
@@ -447,7 +447,7 @@ pub(crate) fn print_jupyter_messages(
         .with_show_fix_status(true)
         .with_show_fix_diff(true)
         .with_show_source(true)
-        .with_unsafe_fixes(UnsafeFixes::Enabled)
+        .with_fix_applicability(Applicability::Unsafe)
         .emit(
             &mut output,
             diagnostics,
@@ -468,7 +468,7 @@ pub(crate) fn print_messages(diagnostics: &[Diagnostic]) -> String {
         .with_show_fix_status(true)
         .with_show_fix_diff(true)
         .with_show_source(true)
-        .with_unsafe_fixes(UnsafeFixes::Enabled)
+        .with_fix_applicability(Applicability::Unsafe)
         .emit(
             &mut output,
             diagnostics,

